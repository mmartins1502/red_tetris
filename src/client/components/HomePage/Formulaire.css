.Form {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 60%;
}

.Input {
  display: flex;
  margin-top: 8px;
}

.Button {
  margin-top: 10px;
  color: #d40e0e;
}
