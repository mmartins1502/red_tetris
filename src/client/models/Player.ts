export type Player = {
  id: string;
  name: string;
  room: string;
};
