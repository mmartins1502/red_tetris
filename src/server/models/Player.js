module.exports = class Player {
  constructor(id, name, room) {
    this.id = id;
    this.name = name;
    this.room = room;
    this.state = false;
  }

  setName(name) {
    this.name = name;
    return this.getName();
  }

  getName() {
    return this.name;
  }

  getSocketId() {
    return this.socketId;
  }
};
