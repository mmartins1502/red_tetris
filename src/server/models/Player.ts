export class Player {
  id: string;
  name: string;
  room: string;
  state: boolean;

  constructor(id: string, name: string, room: string) {
    this.id = id;
    this.name = name;
    this.room = room;
    this.state = false;
  }
}
