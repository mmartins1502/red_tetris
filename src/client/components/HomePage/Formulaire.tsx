import React, { useState, ChangeEvent, FormEvent } from "react";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import "./Formulaire.css";
import { Player } from "../../../Shared/models/Player";

export interface Props {
  onFormValidated: (formData: Player) => void;
}

const Forrmulaire2: React.FC<Props> = (props) => {
  const [state, setState] = useState<any>({
    id: "",
    name: "",
    room: "",
  });

  const changeNameHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setState({
      ...state,
      name: e.target.value
    });
  };

  const changeRoomHandler = (e: ChangeEvent<HTMLInputElement>) => {
    setState({
      ...state,
      room: e.target.value
    });
  };

  const formHandler = (e: FormEvent) => {
    e.preventDefault();
    const formData: any = { ...state };
    props.onFormValidated(formData);
  };

  return (
    <form className="Form" noValidate autoComplete="off">
      <div className="Input">
        <TextField
          inputProps={{
            "data-testid": "name"
          }}
          color="secondary"
          label="Name"
          variant="outlined"
          onChange={(e: ChangeEvent<HTMLInputElement>) => changeNameHandler(e)}
        />
      </div>
      <div className="Input">
        <TextField
          inputProps={{
            "data-testid": "room"
          }}
          color="secondary"
          label="Room"
          variant="outlined"
          onChange={(e: ChangeEvent<HTMLInputElement>) => changeRoomHandler(e)}
        />
      </div>
      <Button
        type="submit"
        data-testid="submit"
        onClick={(e: FormEvent) => formHandler(e)}
        classes={{ label: "Button" }}
      >
        Go to room
      </Button>
    </form>
  );
};

export default Forrmulaire2;
