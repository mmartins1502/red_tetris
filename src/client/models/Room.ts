import { Player } from "./Player";

export type Room = {
  id: string;
  players: Player[];
  inGame: boolean;
  star: Player;
};
