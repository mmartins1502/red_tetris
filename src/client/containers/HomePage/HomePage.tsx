import React, { useState, useEffect, FC } from "react";
import { connect } from "react-redux";

//IMPORT MODELS
import { Player } from "../../models/Player";
import { Room } from "../../models/Room";

//COMPONENTS
import Formulaire from "../../components/HomePage/Formulaire";
import * as actions from "../../store/actions/roomActions";
import { IAppState } from "../../store";
import { Redirect } from "react-router-dom";
import Logo from "../../components/UI/Logo/Logo";

import { Alert } from "@material-ui/lab";
import Slide from "@material-ui/core/Slide";

// CSS & IMG
// const classes = require("./HomePage.module.css");
import "./HomePage.css";

interface IProps {
  room: Room;
  player: Player;
  error: string;
  redirect: boolean;
  onCreatePlayerId: () => void;
  onFormValidated: (formData: Player) => void;
  onRoomNumber: () => void;
}

export const HomePage: FC<IProps> = (props) => {
  // console.log("[HomePage] props", props);

  const { onCreatePlayerId } = props;

  useEffect(() => {
    onCreatePlayerId();
  }, [onCreatePlayerId]);

  const formValidation = (formData: Player) => {
    formData.id = props.player.id;
    props.onFormValidated(formData);
    props.onRoomNumber();

  };

  if (props.redirect && props.room.id && props.player.name && !props.error) {
    return <Redirect to={`/${props.room.id}[${props.player.name}]`} />;
  }

  return (
    <div className="Box">
      <Logo />
      <div className="BoxInput">
        <h2 className="BoxTitle">WELCOME</h2>
        <Formulaire  onFormValidated={(formData) => formValidation(formData)} />
      </div>
      <Slide
        direction="left"
        in={props.error !== ""}
        mountOnEnter
        unmountOnExit
      >
        <Alert style={{ margin: "10px" }} severity="error">
          {props.error}
        </Alert>
      </Slide>
    </div>
  );
};

const mapStateToProps = (store: IAppState) => {
  return {
    playerId: store.socketState.player.id,
    player: store.socketState.player,
    room: store.socketState.room,
    error: store.socketState.error,
    redirect: store.socketState.redirect
  };
};

const mapDispatchToProps = {
  onCreatePlayerId: () => actions.createPlayerId(),
  onFormValidated: (formData: Player) => actions.checkRoom(formData),
  onRoomNumber: () => actions.roomHomeInfos()
};

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
