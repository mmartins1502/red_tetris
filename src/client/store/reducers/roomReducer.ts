import * as actionTypes from "../actions/actionTypes";
import { RoomActions } from "../actions/roomActions";
import { Reducer } from "redux";
import { Player, iPlayer } from '../../../Shared/models/Player';
import { Room, iRoom } from '../../../Shared/models/Room';


export interface iState {
  player: iPlayer;
  room: iRoom; 
  error: string;
  redirect: boolean;
  music: any;
}
const player: iPlayer = new Player("", "", "")
const room: iRoom = new Room("")!
const url = "https://ia600504.us.archive.org/33/items/TetrisThemeMusic/Tetris.mp3";

const initialState: iState = {
  player: player,
  room: room,
  error: "",
  redirect: false,
  music: {
    on: false,
    audio: new Audio(url)
  }
};

// console.log('initialState', initialState.room)

export const roomReducer: Reducer<iState, RoomActions> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case actionTypes.CREATE_PLAYER_ID:
      return {
        ...state,
        player: {
          ...state.player,
          id: action.payload
        }
      };
    case actionTypes.CHECK_ROOM:
      return state;
    case actionTypes.ROOM_AND_PLAYER:
      return {
        ...state,
        player: action.player,
        room: action.room,
        error: action.error,
        redirect: true
      };
    case actionTypes.SETTINGS:
        return {
          ...state,
          room: {
            ...state.room,
            settingsRoom: action.settingsRoom
          }
        };
    case actionTypes.LEAVE_ROOM:
      return state;
    case actionTypes.LEAVE_ROOM_REDUCER:
      return initialState
    case actionTypes.REFRESH_ROOM:
      return {
        ...state,
        room: action.room,
        error: action.error
      };
    case actionTypes.START_GAME:
      return state;
    // case actionTypes.READY:
    //   return state;
    case actionTypes.REFRESH_PLAYER_ASK:
      return state;
    case actionTypes.REFRESH_PLAYER:
        return {
          ...state,
          player: action.player,
          error: action.error
        };
    case actionTypes.SPEED_UP:
      return {
        ...state,
        room: {
          ...state.room,
          speed: action.newSpeed
        }
      };
    case actionTypes.RESET_ROOM:
        return state
    case actionTypes.MUSIC:
      return {
        ...state,
        music: action.music
      }
    case actionTypes.RESET_PLAYER:
      return {
        ...state,
        player: action.player
      }
    default:
      return state;
  }
};
