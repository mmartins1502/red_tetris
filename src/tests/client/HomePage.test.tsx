import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { HomePage } from "../../client/containers/HomePage/HomePage";
import Formulaire from "../../client/components/HomePage/Formulaire";
// import Logo from "../../components/UI/Logo/Logo";
import { Redirect } from "react-router-dom";
import { Alert } from "@material-ui/lab";
import Slide from "@material-ui/core/Slide";
import { Player } from "../../Shared/models/Player";
import { Room } from "../../Shared/models/Room";
import configureStore from "redux-mock-store";
import { iState } from "../../client/store/reducers/roomReducer";


configure({ adapter: new Adapter() });

const configureMockStore = configureStore();

const url = "https://ia600504.us.archive.org/33/items/TetrisThemeMusic/Tetris.mp3";

const initialState: iState = {
  player: new Player("", "", ""),
  room: new Room(""),
  error: "",
  redirect: false,
  music: {
    on: false,
    audio: new Audio(url)
  }
};

const store = configureMockStore(initialState);


let wrapper: any, props: any, useEffect: any;
  props = {
    playerId: "",
    player: new Player("", "", ""),
    room: new Room(""),
    error: "",
    redirect: false,
    onCreatePlayerId: jest.fn(),
    onFormValidated: jest.fn(),
    onRoomNumber: jest.fn()
  }


describe("<HomePage />", () => {
  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f:any) => f());
  };
  beforeEach(() => {
    useEffect = jest.spyOn(React, "useEffect");
    mockUseEffect()
    wrapper = shallow(<HomePage {...props} store={store} />);
  })

  it("should render my home page", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call onCreatePlayerId", () => {
    expect(props.onCreatePlayerId).toHaveBeenCalled()
  })

  it("should call the formValidation function", () => {
    const onFormValidated = jest.fn()
    const formulaire = shallow(<Formulaire onFormValidated={onFormValidated} />);
    formulaire.find({type: "submit"}).simulate('click', { preventDefault: () => {}})
    expect(onFormValidated).toHaveBeenCalled()
    // expect(props.onFormValidated).toHaveBeenCalled()
    expect(props.onRoomNumber).toHaveBeenCalled()
  });

  it("should render my home page with the good error message", () => {
    wrapper.setProps({
      ...props,
      error: "This is my Error Test"
    });
    expect(
      wrapper.containsMatchingElement([
        <h2>WELCOME</h2>,
        <Formulaire onFormValidated={() => {}} />,
        <Slide
          direction="left"
          in={props.error !== ""}
          mountOnEnter
          unmountOnExit
        >
          <Alert style={{ margin: "10px" }} severity="error">
            This is my Error Test
          </Alert>
        </Slide>
      ])
    ).toEqual(true);
  });

  it("should render my home page with <Redirect /> and props.redirect === true", () => {
    wrapper.setProps({
      ...props,
      room: {
        ...props.room,
        id: "testId"
      },
      player: {
        ...props.player,
        name: "testPlayerName"
      },
      redirect: true
    });

    expect(wrapper.find(Redirect)).toHaveLength(1);
    expect(
      wrapper.contains(<Redirect to={`/testId[testPlayerName]`} />)
    ).toEqual(true);
  });
});
