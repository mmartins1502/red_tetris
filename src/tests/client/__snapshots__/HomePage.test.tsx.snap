// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render my home page 1`] = `ShallowWrapper {}`;
